File: src/services/authService.js
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const User = require('../models/User');
const RefreshToken = require('../models/RefreshToken');
const { AppError } = require('../middleware/errorHandler');
const env = require('../config/env');

/**
 * Generate access token
 */
const generateAccessToken = (userId) => {
  return jwt.sign({ id: userId }, env.JWT_SECRET, { expiresIn: env.JWT_EXPIRES_IN });
};

/**
 * Generate refresh token
 */
const generateRefreshToken = (userId) => {
  const { v4: uuidv4 } = require('uuid');
  return jwt.sign({ id: userId, jti: uuidv4() }, env.JWT_REFRESH_SECRET, { expiresIn: env.JWT_REFRESH_EXPIRES_IN });
};

/**
 * Calculate refresh token expiration date
 */
const getRefreshTokenExpiry = () => {
  const expiry = env.JWT_REFRESH_EXPIRES_IN;
  const match = expiry.match(/^(\d+)([smhd])$/);
  if (!match) return new Date(Date.now() + 7 * 24 * 60 * 60 * 1000); // Default 7 days

  const value = parseInt(match[1], 10);
  const unit = match[2];
  const multipliers = { s: 1000, m: 60 * 1000, h: 60 * 60 * 1000, d: 24 * 60 * 60 * 1000 };
  return new Date(Date.now() + value * multipliers[unit]);
};

/**
 * Register a new user
 */
const register = async (userData, ip, userAgent) => {
  const { name, email, password } = userData;

  // Check if user already exists
  const existingUser = await User.findOne({ email: email.toLowerCase() });
  if (existingUser) {
    throw new AppError('An account with this email already exists.', 409, 'EMAIL_EXISTS');
  }

  // Create user
  const user = await User.create({ name, email, password });

  // Generate tokens
  const accessToken = generateAccessToken(user._id);
  const refreshToken = generateRefreshToken(user._id);

  // Store refresh token
  await RefreshToken.create({
    user: user._id,
    token: refreshToken,
    expiresAt: getRefreshTokenExpiry(),
    userAgent: userAgent || '',
    ipAddress: ip || '',
  });

  return { user, accessToken, refreshToken };
};

/**
 * Login user
 */
const login = async (email, password, ip, userAgent) => {
  // Find user with password
  const user = await User.findOne({ email: email.toLowerCase() }).select('+passwordHash');
  if (!user) {
    throw new AppError('Invalid email or password.', 401, 'INVALID_CREDENTIALS');
  }

  if (!user.isActive) {
    throw new AppError('Your account has been deactivated. Contact an administrator.', 403, 'ACCOUNT_DEACTIVATED');
  }

  if (!user.passwordHash) {
    throw new AppError('Account security data is missing. Please register again.', 401, 'MISSING_PASSWORD_HASH');
  }

  // Check password
  const isPasswordValid = await user.comparePassword(password);
  if (!isPasswordValid) {
    throw new AppError('Invalid email or password.', 401, 'INVALID_CREDENTIALS');
  }

  // Update last login
  user.lastLogin = new Date();
  await user.save({ validateBeforeSave: false });

  // Generate tokens
  const accessToken = generateAccessToken(user._id);
  const refreshToken = generateRefreshToken(user._id);

  // Store refresh token
  await RefreshToken.create({
    user: user._id,
    token: refreshToken,
    expiresAt: getRefreshTokenExpiry(),
    userAgent: userAgent || '',
    ipAddress: ip || '',
  });

  return { user, accessToken, refreshToken };
};

/**
 * Refresh access token
 */
const refreshAccessToken = async (refreshTokenValue) => {
  if (!refreshTokenValue) {
    throw new AppError('Refresh token is required.', 400, 'NO_REFRESH_TOKEN');
  }

  // Verify refresh token JWT
  let decoded;
  try {
    decoded = jwt.verify(refreshTokenValue, env.JWT_REFRESH_SECRET);
  } catch {
    throw new AppError('Invalid or expired refresh token.', 401, 'INVALID_REFRESH_TOKEN');
  }

  // Check if token exists in database and is not revoked
  const storedToken = await RefreshToken.findOne({
    token: refreshTokenValue,
    user: decoded.id,
    isRevoked: false,
  });

  if (!storedToken) {
    throw new AppError('Refresh token has been revoked or not found.', 401, 'REVOKED_REFRESH_TOKEN');
  }

  // Check expiration
  if (storedToken.expiresAt < new Date()) {
    throw new AppError('Refresh token has expired.', 401, 'EXPIRED_REFRESH_TOKEN');
  }

  // Revoke old refresh token
  storedToken.isRevoked = true;
  await storedToken.save();

  // Generate new tokens
  const user = await User.findById(decoded.id);
  if (!user || !user.isActive) {
    throw new AppError('User not found or deactivated.', 401, 'USER_UNAVAILABLE');
  }

  const newAccessToken = generateAccessToken(user._id);
  const newRefreshToken = generateRefreshToken(user._id);

  // Store new refresh token
  await RefreshToken.create({
    user: user._id,
    token: newRefreshToken,
    expiresAt: getRefreshTokenExpiry(),
  });

  return { accessToken: newAccessToken, refreshToken: newRefreshToken };
};

/**
 * Logout - revoke refresh token
 */
const logout = async (refreshTokenValue) => {
  if (refreshTokenValue) {
    await RefreshToken.findOneAndUpdate(
      { token: refreshTokenValue },
      { isRevoked: true }
    );
  }
  return { message: 'Logged out successfully' };
};

/**
 * Change password
 */
const changePassword = async (userId, currentPassword, newPassword) => {
  const user = await User.findById(userId).select('+passwordHash');
  if (!user) {
    throw new AppError('User not found.', 404, 'USER_NOT_FOUND');
  }

  if (!user.passwordHash) {
    throw new AppError('Account security data is missing. Please re-register.', 401, 'MISSING_PASSWORD_HASH');
  }

  const isPasswordValid = await user.comparePassword(currentPassword);
  if (!isPasswordValid) {
    throw new AppError('Current password is incorrect.', 401, 'INVALID_PASSWORD');
  }

  user.passwordHash = newPassword;
  await user.save();

  // Revoke all refresh tokens for this user
  await RefreshToken.updateMany({ user: userId }, { isRevoked: true });

  return { message: 'Password changed successfully. Please log in again.' };
};

module.exports = {
  register,
  login,
  refreshAccessToken,
  logout,
  changePassword,
  generateAccessToken,
  generateRefreshToken,
};

File: backend/src/main/java/com/taskflow/service/AuthService.java
package com.taskflow.service;

import com.taskflow.dto.request.AuthRequests;
import com.taskflow.dto.response.ServiceResults;
import com.taskflow.entity.RefreshToken;
import com.taskflow.entity.User;
import com.taskflow.entity.enums.Role;
import com.taskflow.exception.ConflictException;
import com.taskflow.exception.ForbiddenException;
import com.taskflow.exception.ResourceNotFoundException;
import com.taskflow.exception.UnauthorizedException;
import com.taskflow.exception.BadRequestException;
import com.taskflow.repository.RefreshTokenRepository;
import com.taskflow.repository.UserRepository;
import com.taskflow.security.JwtService;
import io.jsonwebtoken.JwtException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Locale;

/** Registration, login, refresh-token rotation, logout and password change. */
@Service
public class AuthService {

    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;

    public AuthService(UserRepository userRepository, RefreshTokenRepository refreshTokenRepository,
                       JwtService jwtService, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.refreshTokenRepository = refreshTokenRepository;
        this.jwtService = jwtService;
        this.passwordEncoder = passwordEncoder;
    }

    @Transactional
    public ServiceResults.AuthResult register(AuthRequests.RegisterRequest request, String ip, String userAgent) {
        String email = request.email().trim().toLowerCase(Locale.ROOT);
        if (userRepository.existsByEmailIgnoreCase(email)) {
            throw new ConflictException("An account with this email already exists.", "EMAIL_EXISTS");
        }

        User user = new User();
        user.setName(request.name().trim());
        user.setEmail(email);
        user.setPasswordHash(passwordEncoder.encode(request.password()));
        user.setRole(Role.USER);
        user.setActive(true);
        user.setAvatar("");
        userRepository.save(user);

        return issueTokens(user, ip, userAgent);
    }

    @Transactional
    public ServiceResults.AuthResult login(String email, String password, String ip, String userAgent) {
        User user = userRepository.findByEmailIgnoreCase(email == null ? "" : email.trim())
                .orElseThrow(() -> new UnauthorizedException("Invalid email or password.", "INVALID_CREDENTIALS"));

        if (!user.isActive()) {
            throw new ForbiddenException("Your account has been deactivated. Contact an administrator.",
                    "ACCOUNT_DEACTIVATED");
        }

        if (user.getPasswordHash() == null || !passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid email or password.", "INVALID_CREDENTIALS");
        }

        user.setLastLogin(LocalDateTime.now());
        userRepository.save(user);

        return issueTokens(user, ip, userAgent);
    }

    @Transactional
    public ServiceResults.TokenPair refreshAccessToken(String refreshTokenValue) {
        if (refreshTokenValue == null || refreshTokenValue.isBlank()) {
            throw new BadRequestException("Refresh token is required.", "NO_REFRESH_TOKEN");
        }

        Long userId;
        try {
            userId = jwtService.parseRefreshToken(refreshTokenValue);
        } catch (JwtException | IllegalArgumentException ex) {
            throw new UnauthorizedException("Invalid or expired refresh token.", "INVALID_REFRESH_TOKEN");
        }

        RefreshToken stored = refreshTokenRepository
                .findByTokenAndUserIdAndRevokedFalse(refreshTokenValue, userId)
                .orElseThrow(() -> new UnauthorizedException("Refresh token has been revoked or not found.",
                        "REVOKED_REFRESH_TOKEN"));

        if (stored.getExpiresAt().isBefore(LocalDateTime.now())) {
            throw new UnauthorizedException("Refresh token has expired.", "EXPIRED_REFRESH_TOKEN");
        }

        stored.setRevoked(true);
        refreshTokenRepository.save(stored);

        User user = userRepository.findById(userId)
                .filter(User::isActive)
                .orElseThrow(() -> new UnauthorizedException("User not found or deactivated.", "USER_UNAVAILABLE"));

        String accessToken = jwtService.generateAccessToken(user.getId());
        String newRefreshToken = jwtService.generateRefreshToken(user.getId());
        persistRefreshToken(user, newRefreshToken, null, null);

        return new ServiceResults.TokenPair(accessToken, newRefreshToken);
    }

    @Transactional
    public String logout(String refreshTokenValue) {
        if (refreshTokenValue != null && !refreshTokenValue.isBlank()) {
            refreshTokenRepository.findByToken(refreshTokenValue).ifPresent(token -> {
                token.setRevoked(true);
                refreshTokenRepository.save(token);
            });
        }
        return "Logged out successfully";
    }

    @Transactional
    public String changePassword(Long userId, String currentPassword, String newPassword) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found.", "USER_NOT_FOUND"));

        if (user.getPasswordHash() == null
                || !passwordEncoder.matches(currentPassword, user.getPasswordHash())) {
            throw new UnauthorizedException("Current password is incorrect.", "INVALID_PASSWORD");
        }

        user.setPasswordHash(passwordEncoder.encode(newPassword));
        userRepository.save(user);
        refreshTokenRepository.revokeAllForUser(userId);

        return "Password changed successfully. Please log in again.";
    }

    private ServiceResults.AuthResult issueTokens(User user, String ip, String userAgent) {
        String accessToken = jwtService.generateAccessToken(user.getId());
        String refreshToken = jwtService.generateRefreshToken(user.getId());
        persistRefreshToken(user, refreshToken, ip, userAgent);
        return new ServiceResults.AuthResult(user, accessToken, refreshToken);
    }

    private void persistRefreshToken(User user, String token, String ip, String userAgent) {
        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setUser(user);
        refreshToken.setToken(token);
        refreshToken.setExpiresAt(LocalDateTime.now().plus(Duration.ofMillis(jwtService.getRefreshExpirationMs())));
        refreshToken.setRevoked(false);
        refreshToken.setIpAddress(ip == null ? "" : ip);
        refreshToken.setUserAgent(userAgent == null ? "" : truncate(userAgent));
        refreshTokenRepository.save(refreshToken);
    }

    private String truncate(String value) {
        return value.length() > 500 ? value.substring(0, 500) : value;
    }
}
